### PR TITLE
experimentation: add configurable termination service

### PR DESCRIPTION
**Pull request by snowp**
Wires up the termination framework as a configurable service that allows configuring an arbitrary set of termination criterias,
including custom ones that can be registered by adding them to the `terminator.CriteriaFactories` map. This allows for
custom pluggable termination criteria that may rely on internal systems (e.g. check for pages using an internal library)
that can hook into the monitoring system provided by the termination framework.

Includes a single termination criteria as an example, which simply terminates experiments after they have run for a configured
amount. 

Rejigs the monitoring class a bit to allow for per config type termination criteria

### CODE REVIEW

File: backend/service/chaos/experimentation/terminator/terminator.go [hunk 1]
```diff
@@ -2,49 +2,119 @@ package terminator
...
+
+	storer, ok := store.(experimentstore.Storer)
+	if !ok {
+		return nil, errors.New("service was not the correct type")
+	}
+
+	terminationCriteria := map[string][]TerminationCriteria{}
+
+	for configType, perConfigTypeConfig := range typedConfig.PerConfigTypeConfiguration {
+		perConfigCriterias := []TerminationCriteria{}
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/terminator/terminator.go:72]:
```suggestion
		perConfigCriteria := []TerminationCriteria{}
```

File: backend/service/chaos/experimentation/terminator/terminator.go [hunk 2]
```diff
@@ -94,13 +164,13 @@ func (m *monitor) Run(ctx context.Context) {
...
 			}
-		}(configType)
+		}(configType, criteria)
 	}
 }
 
 // Iterates over all the provided experiments, spawning a goroutine to montior each experiment that doesn't already have
 // a monitoring routine. Returns a set containing all the active experiment ids for further processing.
-func (m *monitor) monitorNewExperiments(es []*experimentationv1.Experiment, trackedExperiments map[uint64]context.CancelFunc) map[uint64]struct{} {
+func (m *Monitor) monitorNewExperiments(es []*experimentationv1.Experiment, trackedExperiments map[uint64]context.CancelFunc, criterias []TerminationCriteria) map[uint64]struct{} {
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/terminator/terminator.go:173]:
```suggestion
func (m *Monitor) monitorNewExperiments(es []*experimentationv1.Experiment, trackedExperiments map[uint64]context.CancelFunc, criteria []TerminationCriteria) map[uint64]struct{} {
```

File: backend/service/chaos/experimentation/terminator/terminator.go [hunk 3]
```diff
@@ -112,14 +182,14 @@ func (m *monitor) monitorNewExperiments(es []*experimentationv1.Experiment, trac
...
-				m.monitorSingleExperiment(ctx, e)
+				m.monitorSingleExperiment(ctx, e, criterias)
 			}()
 		}
 	}
 	return activeExperiments
 }
 
-func (m *monitor) monitorSingleExperiment(ctx context.Context, e *experimentationv1.Experiment) {
+func (m *Monitor) monitorSingleExperiment(ctx context.Context, e *experimentationv1.Experiment, criterias []TerminationCriteria) {
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/terminator/terminator.go:192]:
```suggestion
func (m *Monitor) monitorSingleExperiment(ctx context.Context, e *experimentationv1.Experiment, criteria []TerminationCriteria) {
```

File: backend/service/chaos/experimentation/terminator/terminator.go [hunk 4]
```diff
@@ -2,49 +2,119 @@ package terminator
...
+	return &Monitor{
+		store:                        storer,
+		terminationCriteriaByTypeUrl: terminationCriteria,
+		outerLoopInterval:            outerLoopInterval,
+		perExperimentCheckInterval:   perExperimentCheckInterval,
+		log:                          logger.Sugar(),
+		activeMonitoringRoutines:     trackingGauge{gauge: scope.Gauge("active_monitoring_routines")},
+		criteriaEvaluationSuccess:    scope.Counter("criteria_success"),
+		criteriaEvaluationFailure:    scope.Counter("criteria_failure"),
+		terminationCount:             scope.Counter("terminations"),
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/terminator/terminator.go:110]: feels weird that some of then have `Count` suffix and some of them don't - not a blocker

File: backend/module/chaos/serverexperimentation/xds/xds_integration_test.go [hunk 5]
```diff
@@ -78,18 +82,32 @@ func TestEnvoyFaultsTimeBasedTermination(t *testing.T) {
 
 	criteria := &testCriteria{}
 
-	terminator := terminator.NewTestMonitor(
-		ts.Storer,
-		[]string{"type.googleapis.com/clutch.chaos.serverexperimentation.v1.HTTPFaultConfig"},
-		[]terminator.TerminationCriteria{criteria},
-		ts.Logger.Sugar(),
-		ts.Scope)
+	terminator.CriteriaFactories["type.googleapis.com/google.protobuf.StringValue"] = &testCriteriaFactory{testCriteria: criteria}
```
Comment (Augustyniak) [inline backend/module/chaos/serverexperimentation/xds/xds_integration_test.go:85]: Now that I look at this - should this test be in `xds_integration_test` file?

Comment (snowp) [reply backend/module/chaos/serverexperimentation/xds/xds_integration_test.go:85]: Up for discussion I guess: it verifies that we actually stop experiments at the xDS level during a termination, but if you think it makes sense to move to another file I can do that.

There are potentially some work that should be done around the integration test framework to better support multiple files, but I can take a look at it

File: backend/service/chaos/experimentation/terminator/terminator.go (hunk 4 above)
Comment (snowp) [reply backend/service/chaos/experimentation/terminator/terminator.go:110]: good point, ill make these consistent

File: backend/module/chaos/serverexperimentation/xds/xds_integration_test.go (hunk 5 above)
Comment (Augustyniak) [reply backend/module/chaos/serverexperimentation/xds/xds_integration_test.go:85]: let's keep it the way it is for now then

File: backend/service/chaos/experimentation/terminator/terminator.go [hunk 6]
```diff
@@ -2,72 +2,142 @@ package terminator
...
+	}
+
+	store, ok := service.Registry[experimentstore.Name]
+	if !ok {
+		return nil, errors.New("could not find experiment store service")
+	}
+
+	storer, ok := store.(experimentstore.Storer)
+	if !ok {
+		return nil, errors.New("service was not the correct type")
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/terminator/terminator.go:66]: super nit
```suggestion
		return nil, errors.New("service was not of the correct type")
```

File: backend/service/chaos/experimentation/terminator/terminator.go [hunk 7]
```diff
@@ -2,72 +2,142 @@ package terminator
...
+
+	terminationCriteria := map[string][]TerminationCriteria{}
+
+	for configType, perConfigTypeConfig := range typedConfig.PerConfigTypeConfiguration {
+		perConfigCriteria := []TerminationCriteria{}
+
+		for _, c := range perConfigTypeConfig.TerminationCriteria {
+			factory, ok := CriteriaFactories[c.TypeUrl]
+			if !ok {
+				return nil, fmt.Errorf("terminator module configured with unknown criteria '%s'", c.TypeUrl)
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/terminator/terminator.go:77]:
```suggestion
				return nil, fmt.Errorf("terminator module configured with unknown criterion '%s'", c.TypeUrl)
```

File: backend/service/chaos/experimentation/terminator/terminator.go [hunk 8]
```diff
@@ -2,72 +2,142 @@ package terminator
...
+			}
+
+			criteria, err := factory.Create(c)
+			if err != nil {
+				return nil, fmt.Errorf("failed to create termination criteria '%s': %s", c.TypeUrl, err)
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/terminator/terminator.go:82]:
```suggestion
				return nil, fmt.Errorf("failed to create termination criterion '%s': %s", c.TypeUrl, err)
```

File: backend/service/chaos/experimentation/terminator/terminator.go [hunk 9]
```diff
@@ -2,72 +2,142 @@ package terminator
...
+	for configType, perConfigTypeConfig := range typedConfig.PerConfigTypeConfiguration {
+		perConfigCriteria := []TerminationCriteria{}
+
+		for _, c := range perConfigTypeConfig.TerminationCriteria {
+			factory, ok := CriteriaFactories[c.TypeUrl]
+			if !ok {
+				return nil, fmt.Errorf("terminator module configured with unknown criteria '%s'", c.TypeUrl)
+			}
+
+			criteria, err := factory.Create(c)
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/terminator/terminator.go:80]:
```suggestion
			criterion, err := factory.Create(c)
```

**Review by Augustyniak — APPROVED** [review]
Nice! left a few nit naming comments - definitely not anything that's big.

File: backend/module/chaos/serverexperimentation/xds/xds_integration_test.go [hunk 10]
```diff
@@ -222,6 +240,14 @@ func awaitExpectedReturnValueForSimpleCall(t *testing.T, e *envoytest.EnvoyHandl
 	return nil
 }
 
+type testCriteriaFactory struct {
+	testCriteria *testCriteria
+}
+
+func (t *testCriteriaFactory) Create(*any.Any) (terminator.TerminationCriteria, error) {
```
Comment (Augustyniak) [inline backend/module/chaos/serverexperimentation/xds/xds_integration_test.go:247]:
```suggestion
func (t *testCriteriaFactory) Create(*any.Any) (terminator.TerminationCriterion, error) {
```

File: backend/service/chaos/experimentation/terminator/max_time_criteria.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+	"github.com/golang/protobuf/ptypes"
+	"github.com/golang/protobuf/ptypes/any"
+	"google.golang.org/protobuf/proto"
+
+	experimentationv1 "github.com/lyft/clutch/backend/api/chaos/experimentation/v1"
+	terminatorv1 "github.com/lyft/clutch/backend/api/config/service/chaos/experimentation/terminator/v1"
+)
+
+type maxTimeTerminationCriteria struct {
```
Comment (Augustyniak) [inline backend/service/chaos/experimentation/terminator/max_time_criteria.go:15]:
```suggestion
type maxTimeTerminationCriterion struct {
```

**Comment by snowp** [comment]
@lyft/clutch-admin Mind giving this a look? Need admin approval to add this as a default service

**Review by danielhochman — APPROVED** [review]

**Pull request merged**